src/mlpack/tests/main_tests/kmeans_test.cpp

```diff
@@ -0,0 +1,414 @@
+      CLI::RestoreSettings(testName);
+  }
+
+  ~KmTestFixture()
+  {
+      // Clear the settings.
+      CLI::ClearSettings();
+  }
+};
+
+void ResetKmSettings()
+{
+  CLI::ClearSettings();
+  CLI::RestoreSettings(testName);
+}
+
+BOOST_FIXTURE_TEST_SUITE(KmeansMainTest, KmTestFixture);
+
+/**
+ * Checking that number of Clusters are non negative
+ */
+BOOST_AUTO_TEST_CASE(NonNegativeClustersTest)
+{
+  arma::mat inputData;
+  if (!data::Load("vc2.csv", inputData))
+    BOOST_FAIL("Unable to load train dataset vc2.csv!");
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("clusters", (int) -1); // Invalid
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+
+/**
+ * Checking that number of Clusters is less than number of points to be clustered
+ */
+BOOST_AUTO_TEST_CASE(PointsLessThanClustersTest)
+{
+  constexpr int N = 10;
+  constexpr int D = 4;
+
+  arma::mat inputData = arma::randu<arma::mat>(N, D);
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("clusters", (int) 11); // Invalid
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+
+/**
+ * Checking that percentage is between 0 and 1 when --refined_start is specified
+*/
+BOOST_AUTO_TEST_CASE(RefinedStartPercentageTest)
+{
+  int c = 2;
+  double P = 2.0;
+  arma::mat inputData;
+  if (!data::Load("vc2.csv", inputData))
+    BOOST_FAIL("Unable to load train dataset vc2.csv!");
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("refined_start", true);
+  SetInputParam("clusters", c);
+  SetInputParam("percentage", std::move(P));     // Invalid
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+
+/**
+ * Checking percentage is non-negative when --refined_start is specified
+ */
+BOOST_AUTO_TEST_CASE(NonNegativePercentageTest)
+{
+  int c = 2;
+  double P = -1.0;
+  arma::mat inputData;
+  if (!data::Load("vc2.csv", inputData))
+    BOOST_FAIL("Unable to load train dataset vc2.csv!");
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("refined_start", true);
+  SetInputParam("clusters", c);
+  SetInputParam("percentage", P);     // Invalid
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+
+/**
+ * Checking that size and dimensionality of prediction is correct.
+ */
+BOOST_AUTO_TEST_CASE(KmClusteringSizeCheck)
+{
+  int c = 2;
+  arma::mat inputData;
+  if (!data::Load("vc2.csv", inputData))
+    BOOST_FAIL("Unable to load train dataset vc2.csv!");
+
+  size_t col = inputData.n_cols;
+  size_t row = inputData.n_rows;
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("clusters", c);
+  SetInputParam("allow_empty_clusters", false);
```

Comment:
It looks like this is a workaround: the program currently fails, saying that either `allow_empty_clusters` or `kill_empty_clusters` must be specified.  We should fix this in `kmeans_main.cpp` by changing the line

```
RequireOnlyOnePassed({ "allow_empty_clusters", "kill_empty_clusters" }, true);
```

to

```
if (CLI::HasParam("allow_empty_clusters") ||
    CLI::HasParam("kill_empty_clusters"))
  RequireOnlyOnePassed({ "allow_empty_clusters", "kill_empty_clusters" }, true);
```

Can you make that change in this PR?  Then we can remove all of the `SetInputParam("allow_empty_clusters", false);` calls.

Reply:
I've made the changes but I've not removed the calls in last test because this workaround is assuring that all the algorithms yield same result. Otherwise some elements prove to be problematic and test won't pass. I think maybe this difference arises due to different empty cluster policies

Reply:
Let's not commit workarounds to the main codebase.  Please remove those, especially because `SetInputParam("allow_empty_clusters", false)` does not do what you think it does, as boolean-typed parameters should only ever be set to true or left unset.